<a href="https://colab.research.google.com/github/I-AM-PRASHANT-VERMA/Almabetter-voyage-analytics-mlops/blob/main/Gender_classification_ML_model/Gender_classification_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

https://github.com/I-AM-PRASHANT-VERMA/Almabetter-voyage-analytics-mlops/tree/main

### 📌 Project Summary – Gender Classification Model 👤🤖

This notebook builds a gender classification model as one part of the Voyage Analytics travel MLOps project. The goal of this model is to predict whether a user is `male` or `female` by using the user's first name as text input. In a travel analytics platform, this kind of model can support user segmentation, personalization, customer profiling, and recommendation features. For example, when a platform understands user groups better, it can study travel behaviour, booking patterns, and service preferences more clearly.

The project uses the `users.csv` file from the travel dataset as the main source because it contains the required `name` and `gender` columns. The related `flights.csv` and `hotels.csv` files are also loaded so the notebook remains connected to the complete travel analytics dataset. The first important step is data cleaning. Only useful records are kept, missing values are removed, gender labels are standardized, and the first name is extracted from the full name. This is important because first names usually carry stronger gender signals than surnames. After cleaning, the dataset becomes balanced enough for fair model training, with nearly equal male and female records.

For modelling, the notebook compares multiple lightweight machine learning classifiers: Logistic Regression, Naive Bayes, Linear SVM, and a tuned Linear SVM. Names are converted into numeric features using TF-IDF character n-grams. This approach is useful for name-based classification because the model can learn small letter patterns, prefixes, suffixes, and common character combinations. The models are evaluated using accuracy, precision, recall, and weighted F1 score. Weighted F1 is used as the final selection metric because it gives a balanced view of classification performance across both classes.

The notebook also includes a proper MLOps workflow. Weights & Biases (W&B) is used to track experiment runs, metrics, parameters, and the final model artifact. After the best model is selected, it is saved as a versioned joblib file and also as a latest deployment copy in Google Drive. This makes the model ready to be promoted into the local MLOps project, where it can be used by the Flask API, Streamlit dashboard, Docker workflow, and later deployment steps. Overall, this notebook is not only a model training file; it is a complete classification workflow from data preparation to experiment tracking and deployment readiness. 🚀


In [ ]:
# Code Cell 001 - Import project libraries

import pandas as pd  # Used for reading and handling dataset tables.
import numpy as np  # Used for numerical operations when needed.
import joblib  # Used for saving and loading the final trained model.
from google.colab import drive  # Used for connecting Google Drive in Colab.
from sklearn.model_selection import train_test_split  # Used for splitting data into train and test sets.
from sklearn.pipeline import Pipeline  # Used for combining TF-IDF and model into one workflow.
from sklearn.feature_extraction.text import TfidfVectorizer  # Used for converting names into text-based numeric features.
from sklearn.linear_model import LogisticRegression  # Used as one lightweight classification model.
from sklearn.naive_bayes import MultinomialNB  # Used as another simple classification model.
from sklearn.svm import LinearSVC  # Used as a strong lightweight text classification model.
from sklearn.metrics import accuracy_score  # Used for checking overall prediction accuracy.
from sklearn.metrics import precision_score  # Used for checking how correct positive predictions are.
from sklearn.metrics import recall_score  # Used for checking how well actual classes are detected.
from sklearn.metrics import f1_score  # Used for checking balanced model performance.
from sklearn.metrics import classification_report  # Used for detailed performance report.
from sklearn.metrics import confusion_matrix  # Used for checking correct and wrong predictions by class.

print("Libraries imported successfully.")  # Confirms that all required libraries are available.


Libraries imported successfully.


In [ ]:
# Code Cell 002 - Mount Google Drive

drive.mount('/content/drive')  # Connects Google Drive with the Colab notebook.

print("Google Drive mounted successfully.")  # Confirms that Drive is ready to use.


Mounted at /content/drive
Google Drive mounted successfully.


In [ ]:
# Code Cell 003 - Load travel datasets

users_path = '/content/drive/MyDrive/Colab Notebooks/# specialization projects/1st project Voyage Analytics: Integrating MLOps in Travel/dataset flight hotel users/users.csv'  # Points to the users dataset in Google Drive.
flights_path = '/content/drive/MyDrive/Colab Notebooks/# specialization projects/1st project Voyage Analytics: Integrating MLOps in Travel/dataset flight hotel users/flights.csv'  # Keeps the related flights file available for project context.
hotels_path = '/content/drive/MyDrive/Colab Notebooks/# specialization projects/1st project Voyage Analytics: Integrating MLOps in Travel/dataset flight hotel users/hotels.csv'  # Keeps the related hotels file available for project context.

users = pd.read_csv(users_path)  # Loads user details used for gender classification.
flights = pd.read_csv(flights_path)  # Loads flight data so the notebook keeps the full travel dataset context.
hotels = pd.read_csv(hotels_path)  # Loads hotel data so the notebook stays aligned with the full project dataset.

print("Users dataset shape:", users.shape)  # Confirms the number of rows and columns in the users table.
print("Flights dataset shape:", flights.shape)  # Confirms the flight dataset is available.
print("Hotels dataset shape:", hotels.shape)  # Confirms the hotel dataset is available.

print("\nUsers dataset preview:")  # Adds a clear label before showing sample rows.
display(users.head())  # Shows the first rows to verify that columns loaded correctly.


Users dataset shape: (1340, 5)
Flights dataset shape: (271888, 10)
Hotels dataset shape: (40552, 8)

Users dataset preview:


,code,company,name,gender,age
0,0,4You,Roy Braun,male,21
1,1,4You,Joseph Holsten,male,37
2,2,4You,Wilma Mcinnis,female,48
3,3,4You,Paula Daniel,female,23
4,4,4You,Patricia Carson,female,44


✅ The dataset loading step was completed successfully. I loaded three files: `users`, `flights`, and `hotels`. The `users` dataset contains 1340 records and 5 columns, while the `flights` and `hotels` datasets are much larger, with 271888 and 40552 records respectively.

For this gender classification task, the `users` dataset is the main dataset because it includes the required `name` and `gender` columns. The preview also shows that the data has useful columns like `code`, `company`, `name`, `gender`, and `age`. Since the model is going to predict gender from names, I will mainly work with the `name` and `gender` columns in the next steps.

In [ ]:
# Code Cell 004 - Clean gender training data

df = users[['name', 'gender']].copy()  # Keeps only the required columns for gender classification.

df = df.dropna(subset=['name', 'gender'])  # Removes rows where name or gender is missing.

df['name'] = df['name'].astype(str).str.strip()  # Converts names into string format and removes extra spaces.

df['gender'] = df['gender'].astype(str).str.lower().str.strip()  # Converts gender values to lowercase and removes extra spaces.

df = df[df['gender'].isin(['male', 'female'])].copy()  # Keeps only male and female classes.

df = df[df['name'].str.len() > 1].copy()  # Removes names that are too short to be useful.

df = df.drop_duplicates(subset=['name', 'gender']).copy()  # Removes duplicate name-gender records.

df['first_name'] = df['name'].str.split().str[0]  # Extracts first name because it is more useful for gender prediction.

df = df.reset_index(drop=True)  # Resets row numbering after cleaning.

print("Cleaned dataset shape:", df.shape)  # Shows the final dataset size after cleaning.

print("\nGender distribution:")  # Prints a heading for class balance.
print(df['gender'].value_counts())  # Shows the number of male and female records.

print("\nCleaned dataset preview:")  # Prints a heading before previewing cleaned data.
display(df.head())  # Displays the first five cleaned rows.


Cleaned dataset shape: (900, 3)

Gender distribution:
gender
male      452
female    448
Name: count, dtype: int64

Cleaned dataset preview:


,name,gender,first_name
0,Roy Braun,male,Roy
1,Joseph Holsten,male,Joseph
2,Wilma Mcinnis,female,Wilma
3,Paula Daniel,female,Paula
4,Patricia Carson,female,Patricia


✅ In this step, I cleaned the users dataset and kept only the columns needed for the gender classification task. After cleaning, the dataset contains **900 records and 3 columns**: **`name`**, **`gender`**, and the new **`first_name`** column.

The important change here is that I created **`first_name`** from the full name. This was done because gender prediction usually depends more on the first name than the surname. For example, in **“Patricia Carson”**, the name **“Patricia”** gives more useful gender information than **“Carson”**.

The class distribution is also almost balanced, with **452 male** records and **448 female** records. This is good for training because the model gets nearly equal examples from both classes. The preview confirms that the new **`first_name`** column was created correctly, like **Roy**, **Joseph**, **Wilma**, **Paula**, and **Patricia**.

In [ ]:
# Code Cell 005 - Create train and test split

X = df['first_name']  # Uses first name as the input feature.

y = df['gender']  # Uses gender as the target label.

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42, stratify=y)  # Splits data into 80% training and 20% testing with balanced classes.

print("Training data size:", X_train.shape[0])  # Shows the number of training records.

print("Testing data size:", X_test.shape[0])  # Shows the number of testing records.

print("\nTraining label distribution:")  # Prints a heading for training class balance.
print(y_train.value_counts())  # Shows male and female counts in training data.

print("\nTesting label distribution:")  # Prints a heading for testing class balance.
print(y_test.value_counts())  # Shows male and female counts in testing data.


Training data size: 720
Testing data size: 180

Training label distribution:
gender
male      362
female    358
Name: count, dtype: int64

Testing label distribution:
gender
female    90
male      90
Name: count, dtype: int64


✅ In this step, I used the **`first_name`** column as the input feature and **`gender`** as the target column. Then I split the data into training and testing sets using an **80/20 split**.

The output shows that **720 records** were used for training and **180 records** were kept for testing. This gives the model enough data to learn from, while still keeping a separate part of the dataset to check how well it performs on unseen data.

The split is also balanced. In the training data, there are **362 male** and **358 female** records, and in the testing data, there are exactly **90 female** and **90 male** records. This is useful because the model will not be tested unfairly on one class more than the other. The **`stratify=y`** part helped maintain this balance.

In [ ]:
# Code Cell 006 - Train baseline classifiers

models = {  # Stores different lightweight models for comparison.
    "Logistic Regression": LogisticRegression(max_iter=1000),  # Uses Logistic Regression for text classification.
    "Naive Bayes": MultinomialNB(),  # Uses Naive Bayes as a simple baseline model.
    "Linear SVM": LinearSVC(),  # Uses Linear SVM as a strong lightweight classifier.
    "Linear SVM Tuned": LinearSVC(C=0.5)  # Uses a slightly tuned SVM version for comparison.
}  # Ends the model dictionary.

results = []  # Stores evaluation scores of each model.

trained_models = {}  # Stores trained model pipelines.

for model_name, model in models.items():  # Runs training for each model one by one.
    pipeline = Pipeline([  # Creates a complete training pipeline.
        ('tfidf', TfidfVectorizer(analyzer='char_wb', ngram_range=(2, 6))),  # Converts first names into character n-gram TF-IDF features.
        ('model', model)  # Adds the classifier after TF-IDF.
    ])  # Ends the pipeline setup.

    pipeline.fit(X_train, y_train)  # Trains the pipeline on training data.

    y_pred = pipeline.predict(X_test)  # Predicts gender on test data.

    accuracy = accuracy_score(y_test, y_pred)  # Calculates accuracy score.

    precision = precision_score(y_test, y_pred, average='weighted')  # Calculates weighted precision score.

    recall = recall_score(y_test, y_pred, average='weighted')  # Calculates weighted recall score.

    f1 = f1_score(y_test, y_pred, average='weighted')  # Calculates weighted F1-score.

    results.append([model_name, accuracy, precision, recall, f1])  # Saves the model scores.

    trained_models[model_name] = pipeline  # Saves the trained pipeline for later selection.

    print("\nModel trained:", model_name)  # Shows which model finished training.
    print("Accuracy:", round(accuracy, 4))  # Shows rounded accuracy.
    print("F1 Score:", round(f1, 4))  # Shows rounded F1-score.



Model trained: Logistic Regression
Accuracy: 0.8722
F1 Score: 0.8721

Model trained: Naive Bayes
Accuracy: 0.8722
F1 Score: 0.8721

Model trained: Linear SVM
Accuracy: 0.9
F1 Score: 0.8998

Model trained: Linear SVM Tuned
Accuracy: 0.9
F1 Score: 0.8998


✅ In this step, I trained **four lightweight machine learning models** and compared how they performed on the test data. The models were **Logistic Regression**, **Naive Bayes**, **Linear SVM**, and a slightly tuned version called **Linear SVM Tuned**.

For feature extraction, I used **TF-IDF with character n-grams** on the **first names**. This means the model learns small character patterns from names, like starting letters, ending letters, and common letter combinations. This is useful for gender classification because names often have repeated patterns.

The output shows a clear improvement compared to the earlier full-name version. **Logistic Regression** and **Naive Bayes** both reached about **87.22% accuracy**, while **Linear SVM** and **Linear SVM Tuned** reached **90% accuracy** with an F1-score around **0.8998**.

This is a strong result for a lightweight machine learning model. There is no problem in the output. The main point is that using **first names** helped the model perform better, and **Linear SVM** is currently the strongest model from this set.

In [ ]:
# Code Cell 007 - Compare model scores

results_df = pd.DataFrame(results, columns=['Model', 'Accuracy', 'Precision', 'Recall', 'F1 Score'])  # Converts all model scores into a table.

results_df = results_df.sort_values(by='F1 Score', ascending=False).reset_index(drop=True)  # Sorts models from best to worst using F1-score.

print("Model comparison table:")  # Prints a heading for the comparison table.

display(results_df)  # Displays the model comparison table.

best_model_name = results_df.loc[0, 'Model']  # Selects the model with the highest F1-score.

best_model = trained_models[best_model_name]  # Gets the trained pipeline of the best model.

print("Best model selected:", best_model_name)  # Displays the selected best model.


Model comparison table:


,Model,Accuracy,Precision,Recall,F1 Score
0,Linear SVM Tuned,0.900000,0.903186,0.900000,0.899802
1,Linear SVM,0.900000,0.903186,0.900000,0.899802
2,Naive Bayes,0.872222,0.873375,0.872222,0.872124
3,Logistic Regression,0.872222,0.873375,0.872222,0.872124


Best model selected: Linear SVM Tuned


✅ In this step, I compared all the trained models in one table using **Accuracy**, **Precision**, **Recall**, and **F1 Score**. I mainly used **F1 Score** for selecting the best model because it gives a balanced view of the model’s performance.

From the output, **Linear SVM Tuned** and **Linear SVM** both gave the same performance, with **90% accuracy** and an **F1 Score of 0.899802**. Since both models performed equally, there is not a big difference between choosing either one.

The code selected **Linear SVM Tuned** as the best model because it appeared at the top after sorting the results. This is not a problem. It just means that the tuned version with **C = 0.5** gave the same result as the normal Linear SVM, so we can continue with **Linear SVM Tuned** as the final model.

Compared to Logistic Regression and Naive Bayes, the SVM models performed better, so selecting **Linear SVM Tuned** for the final `.joblib` file is a reasonable choice.

In [ ]:
# Code Cell 008 - Evaluate the selected classifier

best_predictions = best_model.predict(X_test)  # Predicts the test data using the selected best model.

print("Best Model:", best_model_name)  # Displays the best model name.

print("\nClassification Report:")  # Prints a heading for detailed metrics.
print(classification_report(y_test, best_predictions))  # Shows precision, recall, F1-score, and support.

print("\nConfusion Matrix:")  # Prints a heading for confusion matrix.
print(confusion_matrix(y_test, best_predictions))  # Shows correct and incorrect predictions for both classes.


Best Model: Linear SVM Tuned

Classification Report:
              precision    recall  f1-score   support

      female       0.94      0.86      0.90        90
        male       0.87      0.94      0.90        90

    accuracy                           0.90       180
   macro avg       0.90      0.90      0.90       180
weighted avg       0.90      0.90      0.90       180


Confusion Matrix:
[[77 13]
 [ 5 85]]


✅ In this step, I checked the detailed performance of the selected best model, **Linear SVM Tuned**.

The model achieved **90% accuracy** on the test data, which is a good improvement compared to the earlier full-name version. The test set had **90 female** and **90 male** records, so the evaluation is balanced.

From the report, the model performed well for both classes. For **female**, the precision is **0.94**, which means when the model predicts female, it is usually correct. For **male**, the recall is **0.94**, which means it correctly identifies most of the actual male names.

The confusion matrix gives a clearer picture:

- **77 female names** were correctly predicted as female.
- **13 female names** were wrongly predicted as male.
- **85 male names** were correctly predicted as male.
- **5 male names** were wrongly predicted as female.

So the model is working well overall, but it still makes more mistakes by classifying some **female names as male**. This is not a coding issue; it just shows the model still has some limitation. Since the overall accuracy and F1-score are both around **90%**, this model is a strong final choice for the project.

In [ ]:
# Code Cell 009 - Test sample names

sample_names = ["Roy", "Patricia", "David", "Amit", "Anjali", "Priya", "John"]  # Creates sample first names for testing.

sample_predictions = best_model.predict(sample_names)  # Predicts gender for the sample names.

sample_output = pd.DataFrame({  # Creates a table for sample prediction results.
    "Name": sample_names,  # Stores the sample names.
    "Predicted Gender": sample_predictions  # Stores the predicted gender values.
})  # Ends the sample result table.

print("Sample prediction results:")  # Prints a heading for sample predictions.

display(sample_output)  # Displays the sample prediction table.


Sample prediction results:


,Name,Predicted Gender
0,Roy,male
1,Patricia,female
2,David,male
3,Amit,male
4,Anjali,female
5,Priya,female
6,John,male


✅ In this step, I tested the final selected model with some sample first names to check how it predicts on normal inputs.

The sample results look good. The model predicted **Roy**, **David**, **Amit**, and **John** as **male**, while **Patricia**, **Anjali**, and **Priya** were predicted as **female**. These predictions match the expected gender for these common sample names.

This is also an improvement from the earlier version, where **Amit** was wrongly predicted as female when the model was trained on full names. After switching to **first names**, the prediction for Amit became correct. So this sample test supports the decision to use **first_name** instead of full `name` for training.

### 📊 W&B Experiment Tracking

The earlier cells trained and compared the classifiers inside the notebook. This section repeats the same comparison inside Weights & Biases so every candidate model has a proper experiment run, saved metrics, configuration details, and traceable run history. For this classification task, weighted F1 score is used as the main selection metric because both male and female predictions matter, and the final model should perform well across both classes instead of only optimizing simple accuracy.


In [ ]:
# Code Cell 010 - Track classifier experiments in W&B

!pip install -q wandb

import json
import os
from datetime import datetime
from pathlib import Path

import wandb

WANDB_PROJECT = "voyage-analytics-gender-classification"  # Keeps gender runs separate from flight price runs.
WANDB_ENTITY = "prashant-verma"  # Uses the same W&B account used for the flight price project.

if not wandb.api.api_key:  # Logs in only when the current Colab runtime is not already authenticated.
    wandb.login()  # Opens the W&B login prompt and uses the API key securely.

gender_tracked_models = {}  # Stores the freshly trained pipelines from W&B-tracked runs.
gender_experiment_rows = []  # Stores metrics locally so the best model can be selected after all runs finish.

for model_name, model in models.items():  # Runs one W&B experiment for each candidate classifier.
    run_name = f"{model_name.lower().replace(' ', '_')}_{datetime.now():%Y%m%d_%H%M%S}"  # Gives each run a readable unique name.

    with wandb.init(
        project=WANDB_PROJECT,
        entity=WANDB_ENTITY,
        name=run_name,
        job_type="classification_training",
        config={
            "model_name": model_name,
            "feature_column": "first_name",
            "target_column": "gender",
            "vectorizer": "TfidfVectorizer(char_wb, ngram_range=(2, 6))",
            "train_rows": int(len(X_train)),
            "test_rows": int(len(X_test)),
        },
        tags=["gender-classification", "wandb", "joblib"],
    ) as run:
        pipeline = Pipeline([  # Builds the full text classification pipeline for this run.
            ('tfidf', TfidfVectorizer(analyzer='char_wb', ngram_range=(2, 6))),  # Converts first names into character pattern features.
            ('model', model),  # Adds the candidate classifier after the text features.
        ])

        pipeline.fit(X_train, y_train)  # Trains the candidate model only on the training split.
        predictions = pipeline.predict(X_test)  # Tests the trained model on unseen test names.

        metrics = {  # Collects the classification scores used for comparison.
            "accuracy": accuracy_score(y_test, predictions),
            "precision_weighted": precision_score(y_test, predictions, average="weighted"),
            "recall_weighted": recall_score(y_test, predictions, average="weighted"),
            "f1_weighted": f1_score(y_test, predictions, average="weighted"),
        }

        wandb.log(metrics)  # Sends the numeric scores to the W&B run dashboard.
        run.summary["best_model_candidate"] = model_name  # Saves the candidate name in the run summary.
        run.summary["selection_metric"] = metrics["f1_weighted"]  # Marks F1 as the model selection metric.

        gender_tracked_models[model_name] = pipeline  # Keeps the trained pipeline for final artifact saving.
        gender_experiment_rows.append({"model_name": model_name, "run_id": run.id, "run_name": run.name, **metrics})  # Saves a compact local run record.

gender_results_df = pd.DataFrame(gender_experiment_rows)  # Converts W&B run results into a comparison table.
gender_results_df = gender_results_df.sort_values("f1_weighted", ascending=False).reset_index(drop=True)  # Ranks models by weighted F1 score.

best_gender_model_name = gender_results_df.loc[0, "model_name"]  # Selects the best model name from the ranked table.
best_gender_model = gender_tracked_models[best_gender_model_name]  # Retrieves the trained pipeline for the selected model.
best_gender_metrics = gender_results_df.loc[0].to_dict()  # Stores the winning metrics for metadata and reporting.

display(gender_results_df)  # Shows the final tracked experiment table in the notebook.
print("Best tracked gender model:", best_gender_model_name)  # Prints the selected model in plain text.
print("Best weighted F1:", round(best_gender_metrics["f1_weighted"], 4))  # Prints the main selection score.


/usr/local/lib/python3.12/dist-packages/notebook/notebookapp.py:191: SyntaxWarning: invalid escape sequence '\/'
  | |_| | '_ \/ _` / _` |  _/ -_)
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results


wandb: Enter your choice: 2


wandb: You chose 'Use an existing W&B account'
wandb: Logging into https://api.wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: Create a new API key at: https://wandb.ai/authorize?ref=models
wandb: Store your API key securely and do not share it.


wandb: Paste your API key and hit enter: ··········


wandb: No netrc file found, creating one.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: prashantverma (prashant-verma) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


accuracy,▁
f1_weighted,▁
precision_weighted,▁
recall_weighted,▁
accuracy,0.87222
best_model_candidate,Logistic Regression
f1_weighted,0.87212
precision_weighted,0.87337
recall_weighted,0.87222
selection_metric,0.87212


accuracy,▁
f1_weighted,▁
precision_weighted,▁
recall_weighted,▁
accuracy,0.87222
best_model_candidate,Naive Bayes
f1_weighted,0.87212
precision_weighted,0.87337
recall_weighted,0.87222
selection_metric,0.87212


accuracy,▁
f1_weighted,▁
precision_weighted,▁
recall_weighted,▁
accuracy,0.9
best_model_candidate,Linear SVM
f1_weighted,0.8998
precision_weighted,0.90319
recall_weighted,0.9
selection_metric,0.8998


accuracy,▁
f1_weighted,▁
precision_weighted,▁
recall_weighted,▁
accuracy,0.9
best_model_candidate,Linear SVM Tuned
f1_weighted,0.8998
precision_weighted,0.90319
recall_weighted,0.9
selection_metric,0.8998


,model_name,run_id,run_name,accuracy,precision_weighted,recall_weighted,f1_weighted
0,Linear SVM Tuned,tc9wre5d,linear_svm_tuned_20260530_184902,0.900000,0.903186,0.900000,0.899802
1,Linear SVM,81ogojv9,linear_svm_20260530_184858,0.900000,0.903186,0.900000,0.899802
2,Naive Bayes,mxcb7zr8,naive_bayes_20260530_184853,0.872222,0.873375,0.872222,0.872124
3,Logistic Regression,9jnjvkkd,logistic_regression_20260530_184845,0.872222,0.873375,0.872222,0.872124


Best tracked gender model: Linear SVM Tuned
Best weighted F1: 0.8998


### ✅ What This Result Means

Each classifier is now tracked in W&B with its own run and metrics. The best model is selected using weighted F1 score, which balances precision and recall across both classes. The next cell saves the selected pipeline as a versioned joblib artifact and also creates a `latest` copy for local deployment. This keeps the gender model aligned with the same MLOps pattern used for the flight price and hotel recommendation models.


In [ ]:
# Code Cell 011 - Save the best gender model artifact

GENDER_REGISTRY_DIR = Path("/content/drive/MyDrive/voyage_analytics/model_registry/gender_classification")  # Stores versioned gender model files in Drive.
GENDER_REGISTRY_DIR.mkdir(parents=True, exist_ok=True)  # Creates the registry folder if this is the first run.

timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")  # Creates a timestamp for the versioned model filename.
versioned_model_path = GENDER_REGISTRY_DIR / f"gender_classifier_model_{timestamp}.joblib"  # Stores the immutable model version.
latest_model_path = GENDER_REGISTRY_DIR / "gender_classifier_model_latest.joblib"  # Stores the deployment-friendly latest copy.
metadata_path = GENDER_REGISTRY_DIR / f"gender_classifier_model_{timestamp}_metadata.json"  # Stores metrics and run details for this version.
latest_metadata_path = GENDER_REGISTRY_DIR / "gender_classifier_model_latest_metadata.json"  # Stores the latest metadata beside the latest model.
legacy_model_path = Path("/content/drive/MyDrive/gender_classifier_best_model.joblib")  # Keeps the old filename working for older app code.

joblib.dump(best_gender_model, versioned_model_path)  # Saves the selected model as a timestamped artifact.
joblib.dump(best_gender_model, latest_model_path)  # Saves the same model as the latest deployment artifact.
joblib.dump(best_gender_model, legacy_model_path)  # Saves a compatibility copy using the earlier notebook filename.

metadata = {  # Keeps the model version understandable without opening W&B.
    "model_name": best_gender_model_name,
    "model_type": "gender_classification",
    "feature_column": "first_name",
    "target_column": "gender",
    "selection_metric": "f1_weighted",
    "metrics": best_gender_metrics,
    "created_at": timestamp,
    "versioned_model_path": str(versioned_model_path),
    "latest_model_path": str(latest_model_path),
    "legacy_model_path": str(legacy_model_path),
}

metadata_path.write_text(json.dumps(metadata, indent=2), encoding="utf-8")  # Saves metadata for this exact model version.
latest_metadata_path.write_text(json.dumps(metadata, indent=2), encoding="utf-8")  # Updates the latest metadata file for deployment.

with wandb.init(
    project=WANDB_PROJECT,
    entity=WANDB_ENTITY,
    name=f"register_gender_model_{timestamp}",
    job_type="register_best_model",
    config=metadata,
    tags=["gender-classification", "best-model", "artifact"],
) as run:
    artifact = wandb.Artifact(
        name="gender_classifier_model",
        type="model",
        description="Best W&B-tracked gender classification model for deployment.",
        metadata=metadata,
    )  # Creates a W&B model artifact for version tracking.
    artifact.add_file(str(versioned_model_path))  # Adds the saved joblib model to the artifact.
    artifact.add_file(str(metadata_path))  # Adds the metadata JSON to the artifact.
    run.log_artifact(artifact, aliases=["latest", "best", best_gender_model_name.lower().replace(" ", "_")])  # Publishes artifact aliases.

print("Versioned model saved:", versioned_model_path)  # Shows the exact saved model path.
print("Latest model saved:", latest_model_path)  # Shows the deployment-friendly model path.
print("Metadata saved:", latest_metadata_path)  # Shows where the latest metadata was saved.


Versioned model saved: /content/drive/MyDrive/voyage_analytics/model_registry/gender_classification/gender_classifier_model_20260530_184906.joblib
Latest model saved: /content/drive/MyDrive/voyage_analytics/model_registry/gender_classification/gender_classifier_model_latest.joblib
Metadata saved: /content/drive/MyDrive/voyage_analytics/model_registry/gender_classification/gender_classifier_model_latest_metadata.json


In [ ]:
# Code Cell 012 - Load the saved gender model and check predictions

loaded_model = joblib.load(latest_model_path)  # Loads the latest saved model to confirm the artifact works after saving.

test_names = ["Amit", "Priya", "John"]  # Uses simple names for a quick deployment-style check.
loaded_predictions = loaded_model.predict(test_names)  # Generates predictions from the saved joblib model.

loaded_prediction_df = pd.DataFrame({  # Builds a readable table for the final check.
    "name": test_names,
    "predicted_gender": loaded_predictions,
})

display(loaded_prediction_df)  # Displays the saved-model prediction results.
print("Saved gender model is ready for local app promotion.")  # Confirms the final artifact is usable.


,name,predicted_gender
0,Amit,male
1,Priya,female
2,John,male


Saved gender model is ready for local app promotion.


### ✅ Final Conclusion – Gender Classification Model 👤⚙️

The gender classification workflow was completed successfully as a practical part of the Voyage Analytics MLOps project. The notebook started by loading the travel datasets and then focused on the `users.csv` file because it contains the two most important columns for this task: `name` and `gender`. The data was cleaned carefully by removing missing values, standardizing gender labels, filtering only valid male and female classes, removing duplicate name-gender records, and extracting the first name from each full name. This made the dataset cleaner, simpler, and more suitable for text-based classification.

The modelling approach is simple but effective. Instead of using the full name directly, the model uses the `first_name` column because first names usually carry stronger gender-related patterns. The names are converted into numeric features using TF-IDF character n-grams. This is a good fit for name classification because it allows the model to learn small character patterns such as common endings, starting letters, and repeated letter groups. Multiple models were compared, including Logistic Regression, Naive Bayes, Linear SVM, and tuned Linear SVM. The final selection is based on weighted F1 score because it gives a balanced view of how well the model performs for both male and female classes.

The earlier evaluation showed that the Linear SVM based model performed strongly, with around 90% accuracy and a strong weighted F1 score on the test set. The classification report and confusion matrix also helped explain where the model performs well and where it still makes mistakes. The sample prediction test showed that common names such as Amit, Priya, and John can be predicted correctly after training on first names. This makes the model useful for a project demo and for integration with a user-facing prediction page.

The MLOps part of the notebook is also important. W&B experiment tracking was added so that every candidate model has a visible run, saved configuration, and tracked metrics. The best model is then saved as a versioned joblib artifact and also as a latest deployment copy. A final load test is included to confirm that the saved model can be reused after training. This proves that the notebook is not just doing temporary training inside Colab; it is producing a reusable model artifact that can be copied into the local MLOps project and connected with Flask API, Streamlit UI, Docker, and deployment workflows. Overall, the gender classification notebook is complete, explainable, and ready to support the wider travel analytics system. ✅
